### PR TITLE
Making vtgate add annotations to DML statements.

### PR DESCRIPTION
**Pull request by erzel**
Hi guys,

This change does the following:

+ Adds a package sqlannotations for parsing and annotating SQL DML statement.s
+ Adds vtgate code for annotating DML statements with keyspace id and unfriendly to replication comments. There's still work to do on V3 Execute method.
+ Changes the filtered replication code to parse the new format of annotations.

Please try to pay close attention as I'm new to all of this and likely to have made mistakes. Specifically, I am not sure if I found all the filtered-replication code that needs to change.

Nominating people to review: @sougou @alainjobart

### CODE REVIEW

File: go/vt/binlog/keyrange_filter.go [hunk 1]
```diff
@@ -40,40 +30,23 @@ func KeyRangeFilterFunc(kit key.KeyspaceIdType, keyrange *pb.KeyRange, sendReply
 				log.Warningf("Not forwarding DDL: %s", statement.Sql)
 				continue
 			case proto.BL_DML:
-				keyspaceIndex := strings.LastIndex(statement.Sql, KEYSPACE_ID_COMMENT)
-				if keyspaceIndex == -1 {
+				keyspaceId, unfriendly, err := annotations.ParseSQLAnnotation(string(statement.Sql))
+				if err != nil {
```
Comment (sougou) [inline go/vt/binlog/keyrange_filter.go]: The separate unfriendly return value feels unnecessary. Maybe you could just return it as error, and we can avoid the additional if statement below?

File: go/vt/vtgate/vtgate.go [hunk 2]
```diff
@@ -245,6 +247,12 @@ func (vtg *VTGate) ExecuteKeyspaceIds(ctx context.Context, sql string, bindVaria
 		return errTooManyInFlight
 	}
 
+	if len(keyspaceIds) == 1 {
```
Comment (sougou) [inline go/vt/vtgate/vtgate.go]: I think this if statement can be pushed inside AnnotateKeyspaceIdIfDML. Would be more robust.

**Comment by sougou** [comment]
Do you think you can change v3 also to use the new code? If not, I can take care of it later.
We'll wait for alain also to look at this.

**Comment by enisoc** [comment]
Unless I'm missing something, it doesn't seem necessary to annotate DMLs without keyspace IDs as filtered-replication-unfriendly. When you get to `case proto.BL_DML:` you already know it's a DML. All we should care about now is: If we can find out the keyspace ID, everything's good; if we can't, log an error that we have to skip it.

If I'm right, we can simplify this by only annotating queries when we know the DML (e.g. `ExecuteKeyspaceIds`) and doing nothing when we don't (e.g. `ExecuteKeyRanges`).

File: go/vt/binlog/keyrange_filter.go [hunk 3]
```diff
@@ -5,30 +5,20 @@
...
 
 import (
-	"encoding/base64"
-	"strconv"
-	"strings"
-
 	log "github.com/golang/glog"
 	"github.com/youtube/vitess/go/vt/binlog/proto"
 	"github.com/youtube/vitess/go/vt/key"
+	annotations "github.com/youtube/vitess/go/vt/sqlannotations"
```
Comment (enisoc) [inline go/vt/binlog/keyrange_filter.go]: It seems like you always import `sqlannotations` as `annotations`. We usually try to avoid renaming on import when possible (except protobufs, where it's conventional), so that the package is always referred to by a single canonical name throughout the codebase. My vote is to call it `sqlannotation` (singular) everywhere.

File: go/vt/vtgate/vtgate.go (hunk 2 above)
Comment (alainjobart) [inline go/vt/vtgate/vtgate.go]: Agreed, then the code below can also use that same call.

File: go/vt/vtgate/resolver_test.go [hunk 4]
```diff
@@ -15,12 +15,11 @@ import (
 
 	mproto "github.com/youtube/vitess/go/mysql/proto"
 	"github.com/youtube/vitess/go/vt/key"
+	pb "github.com/youtube/vitess/go/vt/proto/topodata"
```
Comment (alainjobart) [inline go/vt/vtgate/resolver_test.go]: Why this change? I mean, it doesn't matter, just curious. In all files, we define the protobuf import separately, for clarity, as they;re not libraries, but auto-generated, so it feels a bit cleaner.

(I'm not attached to this, I just find it a bit clearer, if others disagree, I'm fine too).

File: go/vt/sqlannotations/sqlannotations.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	}
+	sqlKW = strings.ToLower(sqlKW)
+	if sqlKW == insertDML || sqlKW == updateDML || sqlKW == deleteDML {
+		return true
+	}
+	return false
+}
+
+var (
+	keyspaceIdReplicationRegexp         = regexp.MustCompile("/\\* vtgate:: keyspace_id:([^ ]*) \\*/")
```
Comment (alainjobart) [inline go/vt/sqlannotations/sqlannotations.go]: The regexp here seems a bit overkill, for just a couple string indexes... We commit 5k QPS+, so performance matters.

Would you mind running a small performance experiment to see the CPU usage of a simple string search? Can be in another change...

File: go/vt/binlog/tables_filter.go [hunk 6]
```diff
@@ -11,7 +11,10 @@ import (
 	"github.com/youtube/vitess/go/vt/binlog/proto"
 )
 
-var STREAM_COMMENT = "/* _stream "
+var (
```
Comment (alainjobart) [inline go/vt/binlog/tables_filter.go]: Are you planning on moving the other annotations (done by vttablet) into this library as well? Would be cool I think to have them all in one place.

File: go/vt/sqlannotations/sqlannotations.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
+// Copyright 2015, Google Inc. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package sqlannotations provides functions
+// for annotating DML statements with keyspace-id
+// comments and parsing them. These annotations
+// are used during filtered-replication to route
+// the DML statement to the correct shard.
+package sqlannotations
```
Comment (alainjobart) [inline go/vt/sqlannotations/sqlannotations.go]: Great call to create this library instead of hacking the other more... :)

Comment (alainjobart) [inline go/vt/sqlannotations/sqlannotations.go]: I don't think it's 100% necessary, but maybe unit-testing this library locally would be a bit better. I know coverage is probably good enough for positive cases with the other tests you added, but error cases are probably not covered.

File: go/vt/binlog/keyrange_filter.go [hunk 8]
```diff
@@ -40,40 +30,23 @@ func KeyRangeFilterFunc(kit key.KeyspaceIdType, keyrange *pb.KeyRange, sendReply
...
-					data, err := base64.StdEncoding.DecodeString(textId)
-					if err != nil {
-						updateStreamErrors.Add("KeyRangeStream", 1)
-						log.Errorf("Error parsing keyspace id: %s", statement.Sql)
-						continue
-					}
-					if !key.KeyRangeContains(keyrange, data) {
-						continue
-					}
+				if !key.KeyRangeContains(keyrange, keyspaceId) {
```
Comment (alainjobart) [inline go/vt/binlog/keyrange_filter.go]: This is so much better, to always use []byte. Good call.

File: go/vt/binlog/keyrange_filter.go (hunk 3 above)
Comment (alainjobart) [inline go/vt/binlog/keyrange_filter.go]: +1

File: go/vt/sqlannotations/sqlannotations.go [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+import (
+	"encoding/hex"
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+// AnnotateKeyspaceIdIfDML annotates 'sql' with the given keyspace id
+// if 'sql' is a DML statement; otherwise, it does nothing.
+func AnnotateKeyspaceIdIfDML(keyspaceId []byte, sql *string) {
```
Comment (enisoc) [inline go/vt/sqlannotations/sqlannotations.go]: From outside this package, this always has to be called to as `sqlannotation.AnnotateKeyspaceIdIfDML`. In idiomatic Go, we try to avoid this stuttering (`annotation.Annotate`). Since the package name cannot be omitted, it's always clear already that the function is talking about a SQL annotation.

I suggest:
```
func AddKeyspaceID(sql string, keyspaceID []byte) string
```

This would be called as:
```
sql = sqlannotation.AddKeyspaceID(sql, keyspaceID)
```

My reasoning is:
* By returning the string instead of passing a pointer, it's more clear when the call site is modifying a value - and it becomes possible to use the function when you don't want to modify the original value.
* The reassignment and parameter order resembles an `append()` call, which is a pattern Go users are familiar with.
* By Go naming conventions, the `ID` in `KeyspaceID` should be all-caps. We have some legacy names that are slow to change since they are used in many places, but new names should follow the convention.
* I'm torn on the `IfDML` suffix, but I'd lean towards leaving it out and merely explaining in the docstring (as you did already) that it does nothing for non-DMLs.

**Comment by alainjobart** [comment]
@enisoc wrote:
--
Unless I'm missing something, it doesn't seem necessary to annotate DMLs without keyspace IDs as filtered-replication-unfriendly. When you get to case proto.BL_DML: you already know it's a DML. All we should care about now is: If we can find out the keyspace ID, everything's good; if we can't, log an error that we have to skip it.

If I'm right, we can simplify this by only annotating queries when we know the DML (e.g. ExecuteKeyspaceIds) and doing nothing when we don't (e.g. ExecuteKeyRanges).
--
we wanted to catch the case where applications that weren't supposed to be running were not. In YouTube for instance, chown has to be turned off, if not, we want to know, not just skip statements. The error path has to be re-thought a bit I think, and this is the first step to provide more clarity. I think it's cleaner to see a statement that says: client used the wrong API, instead of seeing a random SQL statement that has no origin and no comment...

**Comment by alainjobart** [comment]
(wow what's up with font size in my previous post... ;) )

@erzel this is a great change, for your first PR to the repo. It has a refactor, great code organization, great unit test coverage, typo fixes... All of this makes me very happy!

File: go/vt/sqlannotations/sqlannotations.go [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+// Returns:
+// If a keyspace-id comment exists
+//   'keyspaceId' is set to the parsed keyspace id, 'unfriendly' is set to 'false',
+//   and err is set to nil.
+// Otherwise, if a filtered-replication-unfriendly comment exists
+//   'keyspaceId' is set to nil, 'unfriendly' is set to true, and err is nil,
+// Otherwise, an error occured (no annotation is considered an error here), and
+//   'keyspaceId' is set to nil, 'unfriendly' is set to false, and err is set
+//   to an error value.
+func ParseSQLAnnotation(sql string) (keyspaceId []byte, unfriendly bool, err error) {
```
Comment (enisoc) [inline go/vt/sqlannotations/sqlannotations.go]: Similar to before, putting `SQLAnnotation` in the name is redundant since it always has to be called as `sqlannotation.ParseSQLAnnotation()`.

I suggest:
```
func ParseKeyspaceID(sql string) ([]byte, error)
```

Reasoning:
* It's not a function for generally parsing annotations (of which we may later add more). It's specifically just looking for the keyspace ID annotation, and the function name makes clear what it's returning.
* As mentioned elsewhere, I believe it's enough to either succeed in finding the keyspace ID, or not. The caller knows if it can achieve what it wants based on that alone.

**Comment by alainjobart** [comment]
Poor Erez you're gonna need hours to read what we wrote!

As a follow-up, the current code that adds the now-deprecated \/\* EMD comment is in google3, in emd.py, so that will need some clean up later.

But the resharding tests in test/... also had to add EMD comment, if you grep there you will find some, that may now be useless if the queries are going through vtgate.

**Comment by enisoc** [comment]
What if we made it say something like: log.Errorf("Skipping DML without keyspace ID annotation: %q. All DMLs during resharding must use Execute or ExecuteKeyspaceIds through VTGate to support filtered replication", ...)

Maybe that's too verbose, but if we can communicate the right thing, it seems like we could simplify the code a lot.

**Comment by erzel** [comment]
Thank you all for the careful look at my change and the great comments.
I'll try go over them tomorrow and address them.

File: go/vt/binlog/keyrange_filter.go [hunk 11]
```diff
@@ -40,40 +30,23 @@ func KeyRangeFilterFunc(kit key.KeyspaceIdType, keyrange *pb.KeyRange, sendReply
...
+						string(statement.Sql), err)
 					continue
 				}
-				idstart := keyspaceIndex + len(KEYSPACE_ID_COMMENT)
-				idend := strings.Index(statement.Sql[idstart:], SPACE)
-				if idend == -1 {
+				if unfriendly {
 					updateStreamErrors.Add("KeyRangeStream", 1)
-					log.Errorf("Error parsing keyspace id: %s", statement.Sql)
+					log.Errorf(
```
Comment (enisoc) [inline go/vt/binlog/keyrange_filter.go]: If we skip DMLs, it's likely that the diff will later fail anyway. Maybe it would be better to halt filtered replication instead of skipping? Alternatively, @alainjobart mentioned that some statements might work ok if they're broadcast to all destination shards. In that case at least there's a chance the diff might come up clean, as opposed to skipping the DML altogether.

Comment (alainjobart) [inline go/vt/binlog/keyrange_filter.go]: Let's move that discussion to a cleanup bug, to figure out what we wanna do then. The behavior of this code as it is now is about the same as before the change, but feels clearer on what's happening, so it's a step forward. So let's submit this and move on. Probably the filtered replication part can be sent to @michael-berlin for later consideration.

File: go/vt/vtgate/resolver_test.go (hunk 4 above)
Comment (erzel) [inline go/vt/vtgate/resolver_test.go]: Oops, I didn't mean to make this change. Reverted.

File: go/vt/binlog/keyrange_filter.go (hunk 11 above)
Comment (erzel) [inline go/vt/binlog/keyrange_filter.go]: SGTM.

File: go/vt/binlog/keyrange_filter.go (hunk 1 above)
Comment (erzel) [inline go/vt/binlog/keyrange_filter.go]: Hmm. I agreed at first, but after thinking more about it, I think we'd like to have a separate case for "unfriendly". We want to know that batch jobs that are not supposed to be running (and generating unfriendly statements) are, rather than just skip statements. I'd also like to change this code to abort the filtered replication, in this case, rather than skipping it. 
What do you think?

File: go/vt/vtgate/vtgate.go (hunk 2 above)
Comment (erzel) [inline go/vt/vtgate/vtgate.go]: Done.

File: go/vt/binlog/keyrange_filter.go (hunk 3 above)
Comment (erzel) [inline go/vt/binlog/keyrange_filter.go]: Done.

File: go/vt/sqlannotations/sqlannotations.go (hunk 7 above)
Comment (erzel) [inline go/vt/sqlannotations/sqlannotations.go]: Right. I was thinking that too. Let me add these.

File: go/vt/binlog/keyrange_filter.go (hunk 8 above)
Comment (erzel) [inline go/vt/binlog/keyrange_filter.go]: Thanks and thanks for making sure it's the right call.

File: go/vt/sqlannotations/sqlannotations.go (hunk 9 above)
Comment (erzel) [inline go/vt/sqlannotations/sqlannotations.go]: I personally prefer a pointer since it's less verbose and I feel like in most call sites the "&" and function name would indicate that the parameter  is changed; however, it's not critical so I changed it to use a return value instead.

I do like the "IfDML syntax since it makes it more explicit to the reader what is actually going
on without having to look in the function comment or code. So I'm leaving that in. Let me know if you want to discuss this further

So I'd like to have:
func AddKeyspaceIDIfDML(sql string, keyspaceID [] byte) string

As per @sougou's comment above, I also have:
func AddIfDML(sql string, keyspaceIDs [][]byte) string
that adds the right comment depending on keyspaceIDs.

File: go/vt/binlog/keyrange_filter.go (hunk 1 above)
Comment (sougou) [inline go/vt/binlog/keyrange_filter.go]: If the action performed is different, then there's a justification for a separate return value.
I've also seen idiomatic uses where they check for a specific error to perform a different action. I think both approaches are reasonable.

Comment (alainjobart) [inline go/vt/binlog/keyrange_filter.go]: Yes the way to do this is to declare a specific constant error in the other file, document in the function that you can return that error in that case, and then when you call the method, check for all three cases with a switch on error (nil, the specific error, and default). I'd go both ways though, I don;t have a preference.

File: go/vt/sqlannotations/sqlannotations.go (hunk 9 above)
Comment (enisoc) [inline go/vt/sqlannotations/sqlannotations.go]: SGTM

**Comment by erzel** [comment]
Hi guys,

Please take another look. I've addressed most of your comments.
Still needs to be done in a forthcoming PR:
+ Make filtered-replication stop when encountering an "unfriendly" comment.
+ As we now annotate using hex format always, I think the KeyspaceIdType parameter in
binlog StreamKeyRange is no longer required. So we need to remove it. Please verify that I'm right here.

@sougou I've changed the V3 API to use the new format for annotations. Can you please go over it and make sure I haven't missed anything?

 Thanks,

 Erez.

File: go/vt/binlog/tables_filter.go (hunk 6 above)
Comment (erzel) [inline go/vt/binlog/tables_filter.go]: Added TODO.

File: go/vt/sqlannotations/sqlannotations.go (hunk 5 above)
Comment (erzel) [inline go/vt/sqlannotations/sqlannotations.go]: I was about to answer that in C++, the RE2 package is pretty efficient and I've seen it used in server code often. However, when I tested this (see the new benchmark), the Go REGEXP package was ~6 times slower than regular strings, so I changed it to strings. Thanks for making me do this.

File: test/utils.py [hunk 12]
```diff
@@ -1120,3 +1121,14 @@ def vtctl_client(self, args):
          '-stderrthreshold', log_level] + args,
         trap_output=True)
     return out
+
+def uint64_to_hex(integer):
+  """Returns the hexadecimal representation of integer treated as a 64-bit unsigned integer.
```
Comment (alainjobart) [inline test/utils.py]: This is only 32 bits here, so its a bit confusing :)

**Comment by alainjobart** [comment]
LGTM, minor nitpick

**Comment by sougou** [comment]
LGTM for v3.

File: test/utils.py (hunk 12 above)
Comment (erzel) [inline test/utils.py]: Of course. Thanks for catching.
